src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/AnsiStringMarshaller.cs

```diff
@@ -9,90 +9,109 @@ namespace System.Runtime.InteropServices.Marshalling
     /// Marshaller for ANSI strings
     /// </summary>
     [CLSCompliant(false)]
-    [CustomTypeMarshaller(typeof(string), BufferSize = 0x100,
-        Features = CustomTypeMarshallerFeatures.UnmanagedResources | CustomTypeMarshallerFeatures.TwoStageMarshalling | CustomTypeMarshallerFeatures.CallerAllocatedBuffer)]
-    public unsafe ref struct AnsiStringMarshaller
+    [CustomMarshaller(typeof(string), MarshalMode.Default, typeof(AnsiStringMarshaller))]
+    [CustomMarshaller(typeof(string), MarshalMode.ManagedToUnmanagedIn, typeof(ManagedToUnmanagedIn))]
+    public static unsafe class AnsiStringMarshaller
     {
-        private byte* _nativeValue;
-        private bool _allocated;
+        /// <summary>
+        /// Convert a string to an unmanaged version.
+        /// </summary>
+        /// <param name="managed">A managed string</param>
+        /// <returns>An unmanaged string</returns>
+        public static byte* ConvertToUnmanaged(string? managed)
+            => (byte*)Marshal.StringToCoTaskMemAnsi(managed);
+
+        /// <summary>
+        /// Convert an unmanaged string to a managed version.
+        /// </summary>
+        /// <param name="unmanaged">An unmanaged string</param>
+        /// <returns>A managed string</returns>
+        public static string? ConvertToManaged(byte* unmanaged)
+            => Marshal.PtrToStringAnsi((nint)unmanaged);
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="AnsiStringMarshaller"/>.
+        /// Free the memory for the unmanaged string.
         /// </summary>
-        /// <param name="str">The string to marshal.</param>
-        public AnsiStringMarshaller(string? str)
-            : this(str, default)
-        { }
+        /// <param name="unmanaged">Memory allocated for the unmanaged string.</param>
+        public static void Free(byte* unmanaged)
+            => Marshal.FreeCoTaskMem((nint)unmanaged);
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="AnsiStringMarshaller"/>.
+        /// Custom marshaller to marshal a managed string as a ANSI unmanaged string.
         /// </summary>
-        /// <param name="str">The string to marshal.</param>
-        /// <param name="buffer">Buffer that may be used for marshalling.</param>
-        /// <remarks>
-        /// The <paramref name="buffer"/> must not be movable - that is, it should not be
-        /// on the managed heap or it should be pinned.
-        /// </remarks>
-        /// <seealso cref="CustomTypeMarshallerFeatures.CallerAllocatedBuffer"/>
-        public AnsiStringMarshaller(string? str, Span<byte> buffer)
+        public ref struct ManagedToUnmanagedIn
         {
-            _allocated = false;
+            /// <summary>
+            /// Requested buffer size for optimized marshalling.
+            /// </summary>
+            public static int BufferSize { get; } = 0x100;
```

Comment:
```suggestion
            public static int BufferSize => 0x100;
```
{ get; } = 0x100; means that the value would be cached in readonly static. We do not need that.

Reply:
I assume I should update all uses of that pattern?

Reply:
(same for other marshallers)